src/ExecBackends/ExecSpace.hpp

```diff
@@ -0,0 +1,450 @@
+  const MEMBACKEND& mem_backend() const
+  {
+    return mb_;
+  }
+
+  const EXECPOLICIES& exec_policies() const
+  {
+    return ep_;
+  }
+  
+  template<typename T, typename I>
+  inline T* alloc_array(const I& n)
+  {
+    return AllocImpl<MEMBACKEND, T, I>::alloc(mb_, n);
+  }
+
+  template<typename T>
+  inline void dealloc_array(T* p)
+  {
+    DeAllocImpl<MEMBACKEND, T>::dealloc(mb_, p);
+  }
+
+  /**
+   * Copy `n` elements of the array `p_src` to the `p_dest` array.
+   * 
+   * @pre `p_src` and `p_dest` should be allocated so that they can hold at least 
+   * `n` elements.
+   * @pre `p_dest` should be managed by the memory backend of `this`.
+   * @pre `p_src` should be managed by the memory backend of `ms`.
+   */
+  template<class MEMSRC, class EXEPOLSRC, typename T, typename I>
+  inline bool copy(T* p_dest, const T* p_src, const I& n, const ExecSpace<MEMSRC,EXEPOLSRC>& ms)
+  {
+    return TransferImpl<MEMBACKEND, EXECPOLICIES, MEMSRC, EXEPOLSRC, T, I>::do_it(p_dest, *this, p_src, ms, n);
+  }
+
+  /**
+   * Copy `n` elements of the array `p_src` to the `p_dest` array.
+   * 
+   * @pre `p_src` and `p_dest` should be allocated so that they can hold at least 
+   * `n` elements.
+   * @pre Both `p_dest` and `p_src` should be managed by the memory backend of `this`.
+   */  
+  template<typename T, typename I>
+  inline bool copy(T* p_dest, const T* p_src, const I& n)
+  {
+    return TransferImpl<MEMBACKEND, EXECPOLICIES, MEMBACKEND, EXECPOLICIES, T, I>::do_it(p_dest, *this, p_src, *this, n);
+  }
+
+private:
+  MEMBACKEND mb_;
+  EXECPOLICIES ep_;
+};
+
+//
+// Internals start here
+//
+  
+/**
+ * Memory allocations should be provided via `AllocImpl` for concrete memory backends. 
+ */
+template<class MEMBACKEND, typename T, typename I>
+struct AllocImpl
+{
+  inline static T* alloc(MEMBACKEND& mb, const I& n)
+  {
+    assert(false && "Specialization for template parameters needs to be provided.");
+    return nullptr;
+  }
+};
+
+/**
+ * Memory deallocations should be provided via `DeAllocImpl` for concrete memory backends.
+ * The size type `I` is not needed by current implementation and defaulted to `void`. 
+ */
+template<class MEMBACKEND, typename T, typename I/*=void*/>
+struct DeAllocImpl
+{
+  inline static void dealloc(MEMBACKEND& mb, T* p)
+  {
+    assert(false && "Specialization for template parameters needs to be provided."); 
+  }
+};
+
+/**
+ * Transfers between memory backends and memory spaces should be provided by specializations
+ * of `TransferImpl` class.
+ */
+template<class MEMDEST, class EXEPOLDEST, class MEMSRC, class EXEPOLSRC, typename T, typename I>
+struct TransferImpl
+{
+  inline static bool do_it(T* p_dest,
+                           ExecSpace<MEMDEST, EXEPOLDEST>& hwb_dest,
+                           const T* p_src,
+                           const ExecSpace<MEMSRC, EXEPOLSRC>& hwb_src,
+                           const I& n)
+  {
+    return false;
+  }
+};
+
+} // end namespace
+
+#endif
```

Comment:
I agree with @pelesh that this should maybe be broken down into separate compilation units for each backend.

I also think that there should perhaps be a `.cpp` file to go along with this one, but that is just prefernce.